include/envoy/network/address.h

```diff
@@ -208,6 +208,7 @@ class Instance {
   virtual const Network::SocketInterface& socketInterface() const PURE;
 };
 
+using InstanceConstPtr = std::unique_ptr<const Instance>;
```

Comment:
Why are we introducing unique_ptr here?

Reply:
Because there are some places where we use unique_ptr of instances, also, unique_ptr can be converted to shared_ptr implicitly, that's the reason why factory method usually return unique_ptr.

Reply:
I am working on adding the main thread assertion and also make Ipv4Instance ctors private. Will ping you when it is ready for another round of review!